pkg/controller/eventing/inmemory/channel/reconcile.go

```diff
@@ -0,0 +1,466 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package channel
+
+import (
+	"context"
+	eventingv1alpha1 "github.com/knative/eventing/pkg/apis/eventing/v1alpha1"
+	"github.com/knative/eventing/pkg/controller"
+	cpcontroller "github.com/knative/eventing/pkg/controller/eventing/inmemory/clusterprovisioner"
+	multichannelfanoutparse "github.com/knative/eventing/pkg/sidecar/configmap/parse"
+	"github.com/knative/eventing/pkg/sidecar/fanout"
+	"github.com/knative/eventing/pkg/sidecar/multichannelfanout"
+	"github.com/knative/eventing/pkg/system"
+	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
+	istiov1alpha3 "github.com/knative/pkg/apis/istio/v1alpha3"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	PortName      = "http"
+	PortNumber    = 80
+	finalizerName = controllerAgentName
+)
+
+type reconciler struct {
+	client   client.Client
+	recorder record.EventRecorder
+	logger   *zap.Logger
+
+	configMapKey client.ObjectKey
+}
+
+// Verify the struct implements reconcile.Reconciler
+var _ reconcile.Reconciler = &reconciler{}
+
+func (r *reconciler) InjectClient(c client.Client) error {
+	r.client = c
+	return nil
+}
+
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// TODO: use this to store the logger and set a deadline
+	ctx := context.TODO()
+	logger := r.logger.With(zap.Any("request", request))
+
+	c := &eventingv1alpha1.Channel{}
+	err := r.client.Get(ctx, request.NamespacedName, c)
+
+	// The Channel may have been deleted since it was added to the workqueue. If so, there is
+	// nothing to be done.
+	if errors.IsNotFound(err) {
+		logger.Info("Could not find Channel", zap.Error(err))
+		return reconcile.Result{}, nil
+	}
+
+	// Any other error should be retried in another reconciliation.
+	if err != nil {
+		logger.Error("Unable to Get Channel", zap.Error(err))
+		return reconcile.Result{}, err
+	}
+
+	// Does this Controller control this Channel?
+	if !r.shouldReconcile(c) {
+		logger.Info("Not reconciling Channel, it is not controlled by this Controller", zap.Any("ref", c.Spec))
+		return reconcile.Result{}, nil
+	}
+	logger.Info("Reconciling Channel")
+
+	err = r.reconcile(ctx, c)
```

Comment:
I'm not a huge fan of it either, what would you suggest to replace it?

Reply:
For what its worth, I dislike the [Subscription controller's way](https://github.com/knative/eventing/blob/7baba63f9488e8937ab704d79e6a96984a9fe4b9/pkg/controller/eventing/subscription/reconcile.go#L62-L70) even more. Mostly because removing the : on line 67 will make it have incorrect behavior.